books/RayTracingInOneWeekend.html

```diff
@@ -2779,6 +2782,239 @@
+        camera cam;
+
+        int image_width       = 100;
+        int image_height      = 100;
+        int samples_per_pixel =  10;
+        int max_depth         =  50;
+
+      private:
+        color ray_color(const ray& r, int depth) {
+            hit_record rec;
+
+            // If we've exceeded the ray bounce limit, no more light is gathered.
+            if (depth <= 0)
+                return color(0,0,0);
+
+            if (world.hit(r, interval(0.001, infinity), rec)) {
+                ray scattered;
+                color attenuation;
+                if (rec.mat_ptr->scatter(r, rec, attenuation, scattered))
+                    return attenuation * ray_color(scattered, depth-1);
+                return color(0,0,0);
+            }
+
+            vec3 unit_direction = unit_vector(r.direction());
+            auto t = 0.5*(unit_direction.y() + 1.0);
+            return (1.0-t)*color(1.0, 1.0, 1.0) + t*color(0.5, 0.7, 1.0);
+        }
+    };
+
+
+    #endif
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    [Listing [scene-initial]: <kbd>[scene.h]</kbd> The new scene class]
+
+</div>
+
+</div class='together'>
+With the new scene class, we trim a lot of code from main.cc. This is what's left:
+
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
+    #include "rtweekend.h"
+
+    #include "camera.h"
+    #include "color.h"
+    #include "hittable_list.h"
+    #include "material.h"
+    #include "scene.h"
+    #include "sphere.h"
+
+    void main() {
+        scene scene_desc;
+
+        const double aspect_ratio = 16.0 / 9.0;
+
+        scene_desc.image_width = 400;
+        scene_desc.image_height = static_cast<int>(scene_desc.image_width / aspect_ratio);
+        scene_desc.samples_per_pixel = 100;
+
+        hittable_list& world = scene_desc.world;
+
+        auto material_ground = make_shared<lambertian>(color(0.8, 0.8, 0.0));
+        auto material_center = make_shared<lambertian>(color(0.1, 0.2, 0.5));
+        auto material_left   = make_shared<dielectric>(1.5);
+        auto material_right  = make_shared<metal>(color(0.8, 0.6, 0.2), 0.0);
+
+        world.add(make_shared<sphere>(point3( 0.0, -100.5, -1.0), 100.0, material_ground));
+        world.add(make_shared<sphere>(point3( 0.0,    0.0, -1.0),   0.5, material_center));
+        world.add(make_shared<sphere>(point3(-1.0,    0.0, -1.0),   0.5, material_left));
+        world.add(make_shared<sphere>(point3(-1.0,    0.0, -1.0),  -0.4, material_left));
+        world.add(make_shared<sphere>(point3( 1.0,    0.0, -1.0),   0.5, material_right));
+
+        scene_desc.render();
+    }
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    [Listing [main-with-scene]: <kdb>[main.cc]</kbd> main.cc using the new scene class]
+
+</div>
+
+Run the new program, you should get the same result as image 16 again.
```

Comment:
same result as before

I  cringe at the idea of specifying an image number. This could break as images as moved around or removed....

Reply:
Changed.

Yeah, this is all a retread of all the different schemes we discussed at length, ending up and the ultimately least-work brute force approach. Issues inevitably come in with the image number, and then we always had to relate the image number from their version back to the current symbolically-named image. Sigh.